### PR TITLE
Task 5 1

### PR DESCRIPTION
**Pull request by ymazhar**

### CODE REVIEW

File: src/api/group/group-router.js [hunk 1]
```diff
@@ -7,20 +7,20 @@ import {
     deleteGroupController,
     addUserGroupController
 } from './controllers/group.controller.js';
-import { validationSchema } from '../../middleware/validateRequest.js';
+import { asyncHandler } from '../../middleware/asynchandler.midleware.js';
 import { groupAddUserSchema, groupIdSchema, groupSchema } from './schema/group.schema.js';
 const groupRouter = Router();
 
-groupRouter.post('/', validationSchema(groupSchema, 'body'), createGroupController);
+groupRouter.post('/', asyncHandler(groupSchema), createGroupController);
```
Comment (BorysLypivets) [inline src/api/group/group-router.js:14]: Need to make few changes:

1. Please, change current line to 
```groupRouter.post('/', createGroupController);```
2. Please go to ```./controllers/group.controller.js``` and change ```createGroupcontroller``` definition to 
`const createGroupController = asyncHandler('Create - Group - Controller', groupSchema, async () => {
... your current handler here
});`
3. remove unused imports of asyncHandler and groupSchema from ```src/api/group/group-router.js```
4. need to do the same for all endpoints in the project

File: src/api/user/controllers/user.controller.js [hunk 2]
```diff
@@ -6,43 +6,75 @@ import {
...
 
-export async function createUserController(req, res) {
-    const body = req.body;
-    const user = await createUser(body);
-
-    if (user.name === 'Error') {
-        res.status(400).send(user.message);
-    } else {
-        res.send(user);
+export async function createUserController(req) {
```
Comment (BorysLypivets) [inline src/api/user/controllers/user.controller.js:9]: please apply changes that described for `src/api/group/group-router.js` for current file

File: src/api/user/controllers/user.controller.js [hunk 3]
```diff
@@ -6,43 +6,75 @@ import {
...
+    try {
```
Comment (BorysLypivets) [inline src/api/user/controllers/user.controller.js:10]: Please remove this try/catch - errors should be handled in your error middleware (need to do the same for all controller handler)

After you applied changes described in the `src/api/group/group-router.js`
you don't need try/catch here, you already have handler().catch() inside asyncHandler function.

File: src/api/user/user-router.js [hunk 4]
```diff
@@ -11,14 +11,14 @@ import { Router } from 'express';
 
 const userRouter = Router();
 
-userRouter.get('/autosuggest?', validationSchema(userAutoSuggestionSchema, 'query'), getAutoSuggestUsersController);
+userRouter.get('/autosuggest?', asyncHandler(userAutoSuggestionSchema, getAutoSuggestUsersController));
```
Comment (BorysLypivets) [inline src/api/user/user-router.js:14]: please use approach described for `src/api/group/group-router.js`

I mean, you should not call asyncHandler inside router files, please move it to controller file

File: src/middleware/asynchandler.midleware.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+export function asyncHandler(schema, handler) {
+    return (req, res, next) => {
+        if (schema.body) {
+            const { error } = schema.body.validate(req.body, {
+                abortEarly: false,
+                allowUnknown: false
+            });
+
+            if (error && error.isJoi) {
+                return next(new SchemaValidationFailedError('Request Body is not valid'));
```
Comment (BorysLypivets) [inline src/middleware/asynchandler.midleware.js:12]: Front End should know which field of the payload failed validation, 
as possible solution your could pass joi error as field of the `SchemaValidationFailedError`

File: src/middleware/error.middleware.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
+import { logger } from './logger.middleware.js';
+
+export const HttpStatusCode = {
+    OK: 200,
+    BAD_REQUEST: 400,
+    NOT_FOUND: 404,
+    INTERNAL_SERVER: 500
+};
+
+export class CustomError extends Error {
```
Comment (BorysLypivets) [inline src/middleware/error.middleware.js:10]: please move definition of Errors to separate file src/lib/error.js

File: src/middleware/error.middleware.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
+import { logger } from './logger.middleware.js';
+
+export const HttpStatusCode = {
```
Comment (BorysLypivets) [inline src/middleware/error.middleware.js:3]: please move definition of status code to separate file src/lib/status.js

File: src/middleware/logger.middleware.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
+import { transports, format, createLogger } from 'winston';
```
Comment (BorysLypivets) [inline src/middleware/logger.middleware.js:1]: Please move this file to src/lib/logger.js 
This file doesn't contain any middleware so it should not be in `middleware` folder

File: src/middleware/asynchandler.midleware.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
+import { SchemaValidationFailedError } from './error.middleware.js';
```
Comment (BorysLypivets) [inline src/middleware/asynchandler.midleware.js:1]: please fix filename asynchandler.mi**d**leware.js -> async-handler.mi**dd**leware.js

File: src/server.js [hunk 10]
```diff
@@ -3,6 +3,8 @@ import express from 'express';
 import { openConnection } from './data-access/db.js';
 import config from './config/index.js';
 import rootRouter from './api/root-router.js';
+import { logger } from './middleware/logger.middleware.js';
+import { logErrorMiddleware, returnError } from './middleware/error.middleware.js';
 
 async function initialize() {
     try {
```
Comment (BorysLypivets) [inline src/server.js:10]: please add `process.on('uncaughtException', ....) ` (Task 5.2 in home work)
on error your should log the exception in with logger and exit process

after your have `process.on('uncaughtException', ....) ` you probably don't need this try/catch any more 
(please check it before your commit changes =) )

File: src/server.js [hunk 11]
```diff
@@ -22,12 +24,19 @@ async function initialize() {
         app.use('/', rootRouter);
 
         app.listen(config.PORT, () => {
-            console.log(`App listening on port ${config.PORT}`);
+            logger.info(`App listening on port ${config.PORT}`);
         });
 
+        process.on('unhandledRejection', error => {
+            throw error;
```
Comment (BorysLypivets) [inline src/server.js:31]: please, change handler to log error, do not throw them from handler

File: src/server.js [hunk 12]
```diff
@@ -22,12 +24,19 @@ async function initialize() {
         app.use('/', rootRouter);
```
Comment (BorysLypivets) [inline src/server.js:24]: look like we missed Task 5.1 on our discussion.

please add one more middleware before `app.use('/', rootRouter);`
this middleware should:
1. use logger to log request that came to app 
2. use `response.on('finish', () => {})` to add handler for response to log response after it finished

**Review by BorysLypivets — CHANGES_REQUESTED** [review]

**Review by BorysLypivets — CHANGES_REQUESTED** [review]

File: src/api/group/service/group.service.js [hunk 13]
```diff
@@ -1,79 +1,58 @@
...
-        if (!group) {
-            return {
-                error: `Cannot find a group with exist ${id} id`
-            };
-        }
-        return group.toJSON();
-    } catch (error) {
-        throw new Error(error);
+    if (!group) {
+        return new GroupNotExistError(`Cannot find a group with exist id: ${id}`);
```
[the remark this answers is not in the document]
Comment (ymazhar) [reply src/api/group/service/group.service.js]: Fixed

File: src/api/user/service/user.service.js [hunk 14]
```diff
@@ -1,78 +1,59 @@
...
+    return users.rows.sort((a, b) => a.login.localeCompare(b.login));
 }
 
 export async function getUser(id) {
-    try {
-        const user = await UserModel.findByPk(id);
-        const isDeleted = user.get('is_deleted');
+    const user = await UserModel.findByPk(id);
+    if (user === null) {
+        return new UserNotExistError(`Can't find user with id: ${id}`);
```
Comment (BorysLypivets) [inline src/api/user/service/user.service.js:52]: please, change `return new` to `throw new`  (1/2)

File: src/api/user/service/user.service.js [hunk 15]
```diff
@@ -1,78 +1,59 @@
...
 export async function createUser(data) {
-    try {
-        const isUserExistInDatabase = await isRecordExist(UserModel, { login: data.login });
+    const isUserExistInDatabase = await isRecordExist(UserModel, { login: data.login });
 
-        if (isUserExistInDatabase) {
-            return new Error('login already exist');
-        }
+    if (isUserExistInDatabase) {
+        return new UserLoginExistError(`${data.login} login already exist`);
```
Comment (BorysLypivets) [inline src/api/user/service/user.service.js:10]: please, change `return new` to `throw new` (2/2)

**Review by BorysLypivets — CHANGES_REQUESTED** [review]

**Review by BorysLypivets — APPROVED** [review]

**Pull request merged**